From 2ff3f8f4b913ad7b5adb36de569d36c25807f04c Mon Sep 17 00:00:00 2001
From: Chad Monroe <chad@monroe.io>
Date: Thu, 6 Nov 2025 14:04:42 -0800
Subject: [PATCH] mt76: support upgrading passive scans to active

on channels with NO_IR or RADAR flags, wait for beacon before sending
probe requests. allows active scanning and WPS on restricted channels
if another AP is already present.

Fixes: c56d6edebc1f ("wifi: mt76: mt7996: use emulated hardware scan support")
Signed-off-by: Chad Monroe <chad@monroe.io>
Tested-by: Piotr Kubik <piotr.kubik@adtran.com>
---
 mt76.h       |    3 +++
 mt7996/mac.c |    3 +++
 scan.c       |   39 ++++++++++++++++++++++++++++++++++++---
 3 files changed, 42 insertions(+), 3 deletions(-)

--- a/mt76.h
+++ b/mt76.h
@@ -1022,6 +1022,8 @@ struct mt76_dev {
 		struct mt76_phy *phy;
 		int chan_idx;
 		bool offchannel;
+		bool beacon_wait;
+		bool beacon_received;
 	} scan;
 
 #ifdef CONFIG_NL80211_TESTMODE
@@ -1608,6 +1610,7 @@ int mt76_get_rate(struct mt76_dev *dev,
 int mt76_hw_scan(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		 struct ieee80211_scan_request *hw_req);
 void mt76_cancel_hw_scan(struct ieee80211_hw *hw, struct ieee80211_vif *vif);
+void mt76_scan_rx_beacon(struct mt76_dev *dev, struct ieee80211_channel *chan);
 void mt76_sw_scan(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		  const u8 *mac);
 void mt76_sw_scan_complete(struct ieee80211_hw *hw,
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -554,6 +554,9 @@ mt7996_mac_fill_rx(struct mt7996_dev *de
 		qos_ctl = FIELD_GET(MT_RXD10_QOS_CTL, v2);
 		seq_ctrl = FIELD_GET(MT_RXD10_SEQ_CTRL, v2);
 
+		if (ieee80211_is_beacon(fc))
+			mt76_scan_rx_beacon(&dev->mt76, mphy->chandef.chan);
+
 		rxd += 4;
 		if ((u8 *)rxd - skb->data >= skb->len)
 			return -EINVAL;
--- a/scan.c
+++ b/scan.c
@@ -79,6 +79,28 @@ out:
 	rcu_read_unlock();
 }
 
+void mt76_scan_rx_beacon(struct mt76_dev *dev, struct ieee80211_channel *chan)
+{
+	struct mt76_phy *phy;
+
+	if (!dev->scan.req)
+		return;
+
+	if (!dev->scan.beacon_wait)
+		return;
+
+	if (!dev->scan.chan || dev->scan.chan->center_freq != chan->center_freq)
+		return;
+
+	phy = dev->scan.phy;
+	if (!phy)
+		return;
+
+	dev->scan.beacon_received = true;
+	ieee80211_queue_delayed_work(phy->hw, &dev->scan_work, 0);
+}
+EXPORT_SYMBOL_GPL(mt76_scan_rx_beacon);
+
 void mt76_scan_work(struct work_struct *work)
 {
 	struct mt76_dev *dev = container_of(work, struct mt76_dev,
@@ -101,8 +123,12 @@ void mt76_scan_work(struct work_struct *
 		goto out;
 	}
 
-	dev->scan.chan = req->channels[dev->scan.chan_idx++];
-	if (phy->main_chandef.chan &&
+	if (!dev->scan.beacon_received) {
+		dev->scan.chan = req->channels[dev->scan.chan_idx++];
+		dev->scan.beacon_wait = false;
+	}
+
+	if (phy->main_chandef.chan && dev->scan.chan &&
 	    phy->main_chandef.chan->center_freq == dev->scan.chan->center_freq)
 		onchannel = true;
 
@@ -114,16 +140,23 @@ void mt76_scan_work(struct work_struct *
 	dev->scan.offchannel = !onchannel;
 	mt76_set_channel(phy, &chandef, !onchannel);
 
-	if (!req->n_ssids ||
-	    chandef.chan->flags & (IEEE80211_CHAN_NO_IR | IEEE80211_CHAN_RADAR))
+	if (!req->n_ssids)
 		goto out;
 
+	if (chandef.chan->flags & (IEEE80211_CHAN_NO_IR | IEEE80211_CHAN_RADAR)) {
+		if (!dev->scan.beacon_received) {
+			dev->scan.beacon_wait = true;
+			goto out;
+		}
+	}
+
 	if (dev->scan.offchannel)
 		duration = HZ / 16; /* ~60 ms */
 	local_bh_disable();
 	for (i = 0; i < req->n_ssids; i++)
 		mt76_scan_send_probe(dev, &req->ssids[i]);
 	local_bh_enable();
+	dev->scan.beacon_received = false;
 
 out:
 	if (dev->scan.chan && dev->scan.offchannel)
